File: src/pipeline.ts
/**
 * Photo-Tiler Pipeline
 *
 * Main orchestration for converting photogrammetry meshes to 3D Tiles 1.1.
 *
 * Pipeline stages:
 * 1. Ingest: Load geometry from various sources
 * 2. Transform: Apply coordinate transformations (unit, axis, georeference)
 * 3. Tile: Build spatial hierarchy and generate GLB tiles
 * 4. Output: Write tileset.json and tile content
 *
 * Supported input types:
 * - glTF/GLB (direct file)
 * - OBJ (direct file, requires units)
 * - PLY (future)
 */

import * as fs from 'fs/promises';
import * as path from 'path';
import { createLogger } from './utils/logger.js';
import { ValidationError } from './utils/errors.js';
import type {
  PipelineConfig,
  ProcessingResult,
  Georeference,
  LinearUnit,
} from './types.js';

const logger = createLogger('pipeline');

/**
 * Run the full photogrammetry-to-3DTiles conversion pipeline
 */
export async function runPipeline(config: PipelineConfig): Promise<ProcessingResult> {
  const startTime = Date.now();

  logger.info({ inputType: config.input.type }, 'Starting Photo-Tiler pipeline');

  // Validate configuration
  validateConfig(config);

  // Ensure output directory exists
  await fs.mkdir(config.output.outputDir, { recursive: true });

  // Stage 1: Ingest
  logger.info('Stage 1: Ingesting source data');
  const { ingest, getIngestionStats } = await import('./ingestion/index.js');

  const ingestionResult = await ingest({
    inputType: config.input.type,
    geometryPath: config.input.geometryPath,
    units: config.input.units,
    workDir: path.join(config.output.outputDir, '.temp'),
  });

  const stats = getIngestionStats(ingestionResult);
  logger.info(
    {
      featureCount: stats.featureCount,
      triangleCount: stats.triangleCount,
      inputUnits: ingestionResult.detectedUnits?.linearUnit,
    },
    'Ingestion complete'
  );

  // Stage 2: Transform
  logger.info('Stage 2: Applying coordinate transforms');
  const {
    computeTransforms,
    transformFeatures,
    transformBounds,
    centerGeometryAtOrigin,
    applyPreTranslationToRootTransform,
  } = await import('./transform/index.js');

  // Use detected unit scale to convert to meters
  const unitScale = ingestionResult.detectedUnits?.toMeters ?? 1.0;

  const transforms = computeTransforms({
    georeference: config.georeference,
    unitScale: unitScale,
    convertYUpToZUp: true,
  });

  logger.info(
    {
      summary: transforms.summary,
      inputUnits: ingestionResult.detectedUnits?.linearUnit,
      unitScale,
    },
    'Computed transforms'
  );

  // Apply local transform to features
  transformFeatures(ingestionResult.features, transforms.localTransform);

  // Transform global bounds
  const transformedBounds = transformBounds(
    ingestionResult.globalBounds,
    transforms.localTransform
  );

  // Center geometry at origin for proper 3D Tiles placement
  // Cesium best practice: tile content should be centered at local origin,
  // with the root transform positioning the tileset in ECEF.
  const centerOffset = centerGeometryAtOrigin(ingestionResult.features, transformedBounds);

  // Apply center offset to root transform so geometry ends up in correct ECEF position
  const finalRootTransform = applyPreTranslationToRootTransform(
    transforms.rootTransform,
    centerOffset
  );

  // Stage 3: Tile
  logger.info('Stage 3: Building tile hierarchy and generating content');
  const { generateTilesetWithProgress } = await import('./tiling/index.js');

  const tilingResult = await generateTilesetWithProgress(
    ingestionResult.features,
    transformedBounds,
    config.output.outputDir,
    {
      tiling: config.tiling,
      draco: config.output.draco,
      rootTransform: finalRootTransform,
      tilesetVersion: '1.0.0',
      materials: config.output.includeTextures ? ingestionResult.materials : undefined,
      textureCompression: config.output.includeTextures
        ? config.output.textureCompression
        : undefined,
      onProgress: (progress) => {
        logger.debug(progress, 'Tiling progress');
      },
    }
  );

  // Stage 4: Validate (optional)
  let validation: ProcessingResult['validation'];

  if (config.output.validate) {
    logger.info('Stage 4: Validating output');
    const { validateTileset } = await import('./tiling/tileset-writer.js');

    const tilesetJson = JSON.parse(
      await fs.readFile(tilingResult.tilesetPath, 'utf-8')
    );

    const validationResult = validateTileset(tilesetJson);
    validation = {
      valid: validationResult.valid,
      errors: validationResult.errors,
      warnings: validationResult.warnings,
    };

    if (!validationResult.valid) {
      logger.warn({ errors: validationResult.errors }, 'Validation failed');
    } else {
      logger.info('Validation passed');
    }
  }

  // Calculate final statistics
  const processingTimeMs = Date.now() - startTime;
  const totalSizeBytes = await calculateOutputSize(config.output.outputDir);

  const result: ProcessingResult = {
    tilesetPath: tilingResult.tilesetPath,
    tileCount: tilingResult.tileCount,
    featureCount: stats.featureCount,
    triangleCount: stats.triangleCount,
    totalSizeBytes,
    processingTimeMs,
    validation,
    inputUnits: ingestionResult.detectedUnits,
    outputInfo: {
      units: 'meters',
      coordinateSystem: 'ECEF',
      epsg: config.georeference?.epsg,
    },
  };

  logger.info(
    {
      tilesetPath: result.tilesetPath,
      tileCount: result.tileCount,
      featureCount: result.featureCount,
      triangleCount: result.triangleCount,
      totalSizeMB: (totalSizeBytes / 1024 / 1024).toFixed(2),
      processingTimeSec: (processingTimeMs / 1000).toFixed(2),
      inputUnits: result.inputUnits?.linearUnit,
      outputUnits: 'meters',
    },
    'Pipeline complete'
  );

  return result;
}

/**
 * Validate pipeline configuration
 */
function validateConfig(config: PipelineConfig): void {
  const errors: string[] = [];

  // Validate input
  if (!config.input.type) {
    errors.push('Input type is required');
  }

  if (config.input.type === 'gltf') {
    if (!config.input.geometryPath) {
      errors.push('Geometry path is required for glTF input');
    }
  }

  if (config.input.type === 'obj') {
    if (!config.input.geometryPath) {
      errors.push('Geometry path is required for OBJ input');
    }
    if (!config.input.units) {
      errors.push('Units are required for OBJ input (mm, cm, m, ft, in)');
    }
  }

  // Validate output
  if (!config.output.outputDir) {
    errors.push('Output directory is required');
  }

  // Validate georeference if provided
  if (config.georeference) {
    if (!config.georeference.epsg) {
      errors.push('EPSG code is required for georeferencing');
    }
    if (
      config.georeference.origin.easting === undefined ||
      config.georeference.origin.northing === undefined
    ) {
      errors.push('Origin easting and northing are required for georeferencing');
    }
  }

  if (errors.length > 0) {
    throw new ValidationError('Invalid pipeline configuration', { errors });
  }
}

/**
 * Calculate total output size
 */
async function calculateOutputSize(outputDir: string): Promise<number> {
  let totalSize = 0;

  async function walkDir(dir: string): Promise<void> {
    const entries = await fs.readdir(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);

      if (entry.isDirectory()) {
        // Skip temp directories
        if (entry.name.startsWith('.')) continue;
        await walkDir(fullPath);
      } else {
        const stat = await fs.stat(fullPath);
        totalSize += stat.size;
      }
    }
  }

  await walkDir(outputDir);
  return totalSize;
}

/**
 * Create a default pipeline configuration for glTF
 */
export function createDefaultConfig(
  inputPath: string,
  outputDir: string,
  georeference?: Georeference,
  units?: LinearUnit
): PipelineConfig {
  const ext = path.extname(inputPath).toLowerCase();

  if (ext !== '.gltf' && ext !== '.glb') {
    throw new ValidationError(`Unsupported file type: ${ext}. Use createObjConfig for OBJ files.`);
  }

  return {
    input: {
      type: 'gltf',
      geometryPath: inputPath,
      ...(units && { units }),
    },
    georeference,
    tiling: {
      maxDepth: 6,
      maxFeaturesPerTile: 500,
      maxTrianglesPerTile: 100000,
      minTileSize: 1.0,
      geometricErrorDecay: 0.5,
    },
    output: {
      outputDir,
      draco: {
        enabled: true,
        compressionLevel: 7,
        quantizePositionBits: 14,
        quantizeNormalBits: 10,
        quantizeTexcoordBits: 12,
      },
      validate: true,
      includeTextures: true,
    },
  };
}

/**
 * Create configuration for OBJ input (requires units)
 */
export function createObjConfig(
  objPath: string,
  units: LinearUnit,
  outputDir: string,
  georeference?: Georeference
): PipelineConfig {
  return {
    input: {
      type: 'obj',
      geometryPath: objPath,
      units,
    },
    georeference,
    tiling: {
      maxDepth: 6,
      maxFeaturesPerTile: 500,
      maxTrianglesPerTile: 100000,
      minTileSize: 1.0,
      geometricErrorDecay: 0.5,
    },
    output: {
      outputDir,
      draco: {
        enabled: true,
        compressionLevel: 7,
        quantizePositionBits: 14,
        quantizeNormalBits: 10,
        quantizeTexcoordBits: 12,
      },
      validate: true,
      includeTextures: true,
    },
  };
}

/**
 * Quick convert function for simple use cases
 */
export async function convert(
  inputPath: string,
  outputDir: string,
  options?: {
    georeference?: Georeference;
    maxTrianglesPerTile?: number;
    maxFeaturesPerTile?: number;
    draco?: boolean;
    units?: LinearUnit;
  }
): Promise<ProcessingResult> {
  const ext = path.extname(inputPath).toLowerCase();

  let config: PipelineConfig;

  if (ext === '.obj') {
    if (!options?.units) {
      throw new ValidationError('Units are required for OBJ files. Specify units in options.');
    }
    config = createObjConfig(inputPath, options.units, outputDir, options?.georeference);
  } else {
    config = createDefaultConfig(inputPath, outputDir, options?.georeference, options?.units);
  }

  if (options?.maxTrianglesPerTile) {
    config.tiling.maxTrianglesPerTile = options.maxTrianglesPerTile;
  }

  if (options?.maxFeaturesPerTile) {
    config.tiling.maxFeaturesPerTile = options.maxFeaturesPerTile;
  }

  if (options?.draco === false) {
    config.output.draco.enabled = false;
  }

  return runPipeline(config);
}

File: src/pipeline.rs
use std::fs;
use std::time::{Duration, Instant};

use gltf::binary::Glb;
use tracing::{info, warn};

use crate::config::PipelineConfig;
use crate::error::{PhotoTilerError, Result};
use crate::ingestion::{self, IngestionResult};
use crate::tiling::{lod, tileset_writer};
use crate::transform::{self, TransformResult};

/// Summary of a completed pipeline run.
#[derive(Debug)]
pub struct ProcessingResult {
    pub tile_count: usize,
    pub duration: Duration,
}

/// Pipeline orchestrator -- drives the four conversion stages.
pub struct Pipeline;

impl Pipeline {
    /// Run the full conversion pipeline.
    pub fn run(config: &PipelineConfig) -> Result<ProcessingResult> {
        let start = Instant::now();

        info!(input = %config.input.display(), "Starting pipeline");

        // Early exits
        if config.show_georef {
            info!("--show-georef: detecting georeferencing information");
            let result = ingestion::ingest(config)?;
            print_georef(&result);
            return Ok(ProcessingResult {
                tile_count: 0,
                duration: start.elapsed(),
            });
        }

        if config.dry_run {
            info!("--dry-run: scanning input and computing transforms");
            let ingestion_result = ingestion::ingest(config)?;
            let transform_result = transform::transform(config, &ingestion_result)?;
            print_dry_run_summary(&ingestion_result, &transform_result);
            return Ok(ProcessingResult {
                tile_count: 0,
                duration: start.elapsed(),
            });
        }

        // Full pipeline
        info!("Stage 1/4: Ingestion");
        let ingestion_result = ingestion::ingest(config)?;

        info!("Stage 2/4: Transform");
        let transform_result = transform::transform(config, &ingestion_result)?;
        print_transform_summary(&transform_result);

        info!("Stage 3/4: Tiling");
        fs::create_dir_all(&config.output).map_err(|e| {
            PhotoTilerError::Output(format!(
                "Failed to create output directory {}: {e}",
                config.output.display()
            ))
        })?;
        let tile_count = Self::tile(config, transform_result)?;

        if config.validate {
            info!("Stage 4/4: Validation");
            Self::validate(config)?;
        }

        let duration = start.elapsed();
        info!(tiles = tile_count, elapsed = ?duration, "Pipeline complete");

        Ok(ProcessingResult {
            tile_count,
            duration,
        })
    }

    fn tile(config: &PipelineConfig, transform_result: TransformResult) -> Result<usize> {
        let max_lod_levels = 1;

        // Destructure to take ownership of fields individually
        let TransformResult {
            meshes,
            bounds,
            materials,
            root_transform,
        } = transform_result;

        let mesh_count = meshes.len();

        // Move meshes into LOD generation (no extra copies)
        let lod_chains: Vec<_> = meshes
            .into_iter()
            .enumerate()
            .map(|(i, mesh)| {
                info!(
                    mesh = i,
                    vertices = mesh.vertex_count(),
                    triangles = mesh.triangle_count(),
                    "Generating LOD chain"
                );

                let chain = lod::generate_lod_chain(mesh, &bounds, max_lod_levels);

                for level in &chain.levels {
                    info!(
                        mesh = i,
                        lod = level.level,
                        triangles = level.mesh.triangle_count(),
                        geometric_error = level.geometric_error,
                        "LOD level"
                    );
                }

                chain
            })
            .collect();

        let total_lod_levels: usize = lod_chains.iter().map(|c| c.levels.len()).sum();
        info!(
            meshes = mesh_count,
            total_lod_levels,
            "LOD generation complete"
        );

        // Build tile hierarchy and write GLBs eagerly to disk
        info!("Building tile hierarchy");
        let tileset_output = tileset_writer::build_tileset(
            lod_chains,
            &bounds,
            &config.tiling,
            &materials,
            &config.texture,
            &config.output,
        );

        // Write tileset.json (GLBs already on disk)
        info!(output = %config.output.display(), "Writing tileset.json");
        let tile_count =
            tileset_writer::write_tileset(&tileset_output, &root_transform, &config.output)?;

        Ok(tile_count)
    }

    fn validate(config: &PipelineConfig) -> Result<()> {
        let out_dir = &config.output;

        // 1. tileset.json must exist and be valid JSON
        let tileset_path = out_dir.join("tileset.json");
        let json_str = fs::read_to_string(&tileset_path).map_err(|e| {
            PhotoTilerError::Validation(format!(
                "Cannot read tileset.json at {}: {e}",
                tileset_path.display()
            ))
        })?;

        let tileset: serde_json::Value = serde_json::from_str(&json_str).map_err(|e| {
            PhotoTilerError::Validation(format!("tileset.json is not valid JSON: {e}"))
        })?;

        // 2. Required top-level fields
        let asset = tileset
            .get("asset")
            .ok_or_else(|| PhotoTilerError::Validation("Missing 'asset' field".into()))?;
        let version = asset
            .get("version")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        if version != "1.1" {
            return Err(PhotoTilerError::Validation(format!(
                "Expected asset.version '1.1', got '{version}'"
            )));
        }

        let root = tileset
            .get("root")
            .ok_or_else(|| PhotoTilerError::Validation("Missing 'root' tile".into()))?;

        // 3. Walk tile tree: validate each tile
        let mut tile_count = 0;
        let mut glb_count = 0;
        let mut errors = Vec::new();
        validate_tile(root, out_dir, None, &mut tile_count, &mut glb_count, &mut errors);

        for err in &errors {
            warn!("Validation: {err}");
        }

        if errors.is_empty() {
            info!(tiles = tile_count, glbs = glb_count, "Validation passed");
        } else {
            return Err(PhotoTilerError::Validation(format!(
                "{} issues found: {}",
                errors.len(),
                errors.first().unwrap()
            )));
        }

        Ok(())
    }
}

/// Recursively validate a tile node from tileset.json.
fn validate_tile(
    tile: &serde_json::Value,
    out_dir: &std::path::Path,
    parent_error: Option<f64>,
    tile_count: &mut usize,
    glb_count: &mut usize,
    errors: &mut Vec<String>,
) {
    *tile_count += 1;

    // Bounding volume must exist
    if tile.get("boundingVolume").is_none() {
        errors.push(format!("Tile {tile_count}: missing boundingVolume"));
    }

    // Geometric error must be non-negative
    let geo_error = tile
        .get("geometricError")
        .and_then(|v| v.as_f64())
        .unwrap_or(-1.0);
    if geo_error < 0.0 {
        errors.push(format!("Tile {tile_count}: invalid geometricError {geo_error}"));
    }

    // Geometric error should not exceed parent's
    if let Some(parent_err) = parent_error {
        if geo_error > parent_err + 1e-6 {
            errors.push(format!(
                "Tile {tile_count}: geometricError {geo_error} > parent {parent_err}"
            ));
        }
    }

    // If tile has content, verify the GLB file
    if let Some(content) = tile.get("content") {
        if let Some(uri) = content.get("uri").and_then(|u| u.as_str()) {
            let glb_path = out_dir.join(uri);
            if !glb_path.exists() {
                errors.push(format!("Tile {tile_count}: GLB not found: {uri}"));
            } else {
                *glb_count += 1;
                // Try to parse the GLB
                match fs::read(&glb_path) {
                    Ok(data) => {
                        if Glb::from_slice(&data).is_err() {
                            errors.push(format!("Tile {tile_count}: GLB not parseable: {uri}"));
                        }
                    }
                    Err(e) => {
                        errors.push(format!("Tile {tile_count}: cannot read {uri}: {e}"));
                    }
                }
            }
        }
    }

    // Recurse into children
    if let Some(children) = tile.get("children").and_then(|c| c.as_array()) {
        for child in children {
            validate_tile(child, out_dir, Some(geo_error), tile_count, glb_count, errors);
        }
    }
}

/// Print georeferencing information and exit.
fn print_georef(result: &IngestionResult) {
    println!("=== Georeferencing ===");
    match &result.georeference {
        Some(geo) => {
            println!("  EPSG:      {}", geo.epsg);
            println!("  Easting:   {:.3}", geo.easting);
            println!("  Northing:  {:.3}", geo.northing);
            println!("  Elevation: {:.3}", geo.elevation);
            println!("  True North:{:.1}°", geo.true_north);
        }
        None => {
            println!("  No georeference detected.");
            println!("  Use --epsg, --offset-file, or --metadata-xml to specify.");
        }
    }
}

/// Print transform summary: bounding box and root transform.
fn print_transform_summary(result: &TransformResult) {
    let bb = &result.bounds;
    println!("=== Transform ===");
    println!(
        "  Bounding box: ({:.3}, {:.3}, {:.3}) → ({:.3}, {:.3}, {:.3})",
        bb.min[0], bb.min[1], bb.min[2], bb.max[0], bb.max[1], bb.max[2]
    );
    println!("  Diagonal:     {:.3} m", bb.diagonal());

    let rt = &result.root_transform;
    let is_identity = rt[0] == 1.0
        && rt[5] == 1.0
        && rt[10] == 1.0
        && rt[15] == 1.0
        && rt[12] == 0.0
        && rt[13] == 0.0
        && rt[14] == 0.0;

    if is_identity {
        println!("  Root transform: identity (local coordinates)");
    } else {
        println!(
            "  Root transform: ECEF ({:.1}, {:.1}, {:.1})",
            rt[12], rt[13], rt[14]
        );
    }
}

/// Print dry-run summary with mesh stats, georeferencing, and transform info.
fn print_dry_run_summary(ingestion: &IngestionResult, transform: &TransformResult) {
    let stats = &ingestion.stats;
    println!("=== Dry Run Summary ===");
    println!("  Format:    {}", stats.input_format);
    println!("  Meshes:    {}", stats.total_meshes);
    println!("  Vertices:  {}", stats.total_vertices);
    println!("  Triangles: {}", stats.total_triangles);
    println!("  Normals:   {}", if stats.has_normals { "yes" } else { "no" });
    println!("  UVs:       {}", if stats.has_uvs { "yes" } else { "no" });
    println!("  Colors:    {}", if stats.has_colors { "yes" } else { "no" });
    println!("  Materials: {}", stats.material_count);
    println!("  Textures:  {}", stats.texture_count);
    println!();
    print_georef(ingestion);
    println!();
    print_transform_summary(transform);
}
